libtasn1/doc/asn1Parser.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.3.
.TH ASN1PARSER "1" "July 2016" "asn1Parser (libtasn1) 4.9" "User Commands"
.SH NAME
asn1Parser \- ASN.1 syntax tree generator for libtasn1
.SH SYNOPSIS
.B asn1Parser
[\fI\,OPTION\/\fR] \fI\,FILE\/\fR
.SH DESCRIPTION
Read FILE with ASN.1 definitions and generate
a C array that is used with libtasn1 functions.
.PP
Mandatory arguments to long options are mandatory for short options too.
.TP
\fB\-c\fR, \fB\-\-check\fR
checks the syntax only
.TP
\fB\-o\fR, \fB\-\-output\fR=\fI\,FILE\/\fR
output file
.TP
\fB\-n\fR, \fB\-\-name\fR=\fI\,NAME\/\fR
array name
.TP
\fB\-h\fR, \fB\-\-help\fR
display this help and exit
.TP
\fB\-v\fR, \fB\-\-version\fR
output version information and exit
.SH AUTHOR
Written by Fabio Fiorina.
.SH "REPORTING BUGS"
Report bugs to: help\-libtasn1@gnu.org
.br
GNU Libtasn1 home page: <http://www.gnu.org/software/libtasn1/>
.br
General help using GNU software: <http://www.gnu.org/gethelp/>
.SH COPYRIGHT
Copyright \(co 2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
The full documentation for
.B asn1Parser
is maintained as a Texinfo manual.  If the
.B info
and
.B asn1Parser
programs are properly installed at your site, the command
.IP
.B info libtasn1
.PP
should give you access to the complete manual.


NEWS
Changes between 0.3.0 and 0.3.1:
--------------------------------

  * Fix crash when used from multipe threads
  * Fix nl_langinfo detection on Android
  * Fix build for winstore


Changes between 0.2.7 and 0.3.0:
--------------------------------

  * Implement SMB SESSION LOGOFF (0x74)
  * Add smb_stat_list_next() to reduce directory traversal speed
  * Fix guest login
  * Fix compilation when using compatibility layer
  * Fix crash against malformed packet in smb_fstat
  * Fix too small buffer in smb_session_login_spnego, leading to failure
  * Fix payload for smb_negotiate for arm5te
  * Fix misc crashes


Changes between 0.2.6 and 0.2.7:
--------------------------------

  * Fix build when lacking getifaddrs


Changes between 0.2.5 and 0.2.6:
--------------------------------

  * Fix connection with a smb server listening on the port 139
  * Fix compilation and struct packing for MSVC
  * Fix sample code
  * Fix Windows broadcast addresses on multiple interfaces
  * Fix return status on smb_fstat() call


Changes between 0.2.4 and 0.2.5:
--------------------------------

 * Add cancellation support without pipe()
 * Continued port to Windows
 * Fix teardown crash and wrong asserts
 * Fix potential crash
 * Minor build system cleanup


Changes between 0.2.3 and 0.2.4:
--------------------------------

 * Fix compilation for Android
 * Continued port to Windows


Changes between 0.2.2 and 0.2.3:
--------------------------------

 * Partial port to Windows (Win32 and WinRT/UWP)
 * Fix logged state after smb_session_login_ntlm


Changes between 0.2.1 and 0.2.2:
--------------------------------

 * Disable DEBUG mode by default
 * Stability improvements


Changes between 0.2.0 and 0.2.1:
--------------------------------

 * Fix seek with file > 2GB on Android
 * smb_fread can skip bytes


Changes between 0.1.0 and 0.2.0:
--------------------------------

 * Add files and folder deletion with smb_directory_rm and smb_file_rm
 * Add folder creation with smb_directory_create
 * Add files creation with smb_fwrite
 * Add files and folder move/rename with smb_file_mv
 * Fix session state after spnego authentication
 * Fix return values of all public functions: return a negative number in case
   of error, and 0 or a positive number in case of success.
 * Add smb_session_get_nt_status: get the last NT_STATUS
 * Only export needed symbols


Changes between 0.0.7 and 0.1.0:
--------------------------------

 * Fix shares listing when server returns a tid equal to 0
   Notably TimeCapsule and AirPort Express/Extreme servers
 * Fix crashes and wrong assertions
 * Improve BSD compatibility


Changes between 0.0.6 and 0.0.7:
--------------------------------

 * Rework folders opening with large number of files (FIND_NEXT)
 * Fix for reading from OS X SMB Sharing
 * Fix build for old glibc versions
 * Fix support for large files (fix for >1GiB and for >4GiB)
 * Fix potential deadlock when server disconnects


Changes between 0.0.5 and 0.0.6:
--------------------------------

 * Fix iOS compilation
 * Fix regression on smb_buffer_alloca that introduced login and discovery
   issues
 * Fix double-free crash


Changes between 0.0.4 and 0.0.5:
--------------------------------

 * Rework netbios_ns timeouts
 * Expose less APIs and defines
 * Fix support for Darwin OS missing clock_gettime()


Changes between 0.0.3 and 0.0.4:
--------------------------------

 * Improve discovery that is now done in a separate thread and lot more robust
 * API: Add netbios_ns_discover_start and netbios_ns_discover_stop to start and
   stop a discovery task
 * API: Remove netbios_ns_discover, netbios_ns_entry_count, netbios_ns_entry_at,
   netbios_ns_clear, netbios_ns_abort. Use the new discovers calls instead
 * Fix smb_fopen failing to open long path


Changes between 0.0.2 and 0.0.3:
--------------------------------

 * Port to Android
 * Simplification of the NetBios discovery API
 * New abort mechanism when discovery: netbios_ns_abort
 * Numerous cleanups, include headers and API
 * Provide a replacement for strlcpy
 * Fix numerous warnings, build issues and leaks
 * Support SMBX servers


Changes between 0.0.1 and 0.0.2:
--------------------------------

 * Fix various build failures


First release 0.0.1:
--------------------

 * This is the first basic release.
 * It should provide the basics for this library, but is not ready yet for
   prime-time


libtasn1/src/asn1Parser.c
/* asn1Parser.c -- program to parse a file with ASN1 definitions
 * Copyright (C) 2002-2014 Free Software Foundation, Inc.
 *
 * This file is part of LIBTASN1.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <config.h>

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>

#include <libtasn1.h>

#include <progname.h>
#include <version-etc.h>

/* This feature is available in gcc versions 2.5 and later.  */
#if __GNUC__ < 2 || (__GNUC__ == 2 && __GNUC_MINOR__ < 5)
#define ATTR_NO_RETRUN
#else
#define ATTR_NO_RETRUN __attribute__ ((__noreturn__))
#endif

ATTR_NO_RETRUN static void
usage (int status)
{
  if (status != EXIT_SUCCESS)
    fprintf (stderr, "Try `%s --help' for more information.\n", program_name);
  else
    {
      printf ("\
Usage: %s [OPTION] FILE\n", program_name);
      printf ("\
Read FILE with ASN.1 definitions and generate\n\
a C array that is used with libtasn1 functions.\n\
\n");
      printf ("\
Mandatory arguments to long options are mandatory for short options too.\n\
  -c, --check           checks the syntax only\n\
  -o, --output=FILE     output file\n\
  -n, --name=NAME       array name\n\
  -h, --help            display this help and exit\n\
  -v, --version         output version information and exit\n");
      emit_bug_reporting_address ();
    }
  exit (status);
}

int
main (int argc, char *argv[])
{
  static const struct option long_options[] = {
    {"help", no_argument, 0, 'h'},
    {"version", no_argument, 0, 'v'},
    {"check", no_argument, 0, 'c'},
    {"output", required_argument, 0, 'o'},
    {"name", required_argument, 0, 'n'},
    {0, 0, 0, 0}
  };
  int option_index = 0;
  int option_result;
  char *outputFileName = NULL;
  char *inputFileName = NULL;
  char *vectorName = NULL;
  int checkSyntaxOnly = 0;
  asn1_node pointer = NULL;
  char errorDescription[ASN1_MAX_ERROR_DESCRIPTION_SIZE];
  int parse_result = ASN1_SUCCESS;

  set_program_name (argv[0]);

  opterr = 0;			/* disable error messages from getopt */

  while (1)
    {

      option_result =
	getopt_long (argc, argv, "hvco:n:", long_options, &option_index);

      if (option_result == -1)
	break;

      switch (option_result)
	{
	case 0:
	  fprintf (stderr, "option %s", long_options[option_index].name);
	  if (optarg)
	    fprintf (stderr, " with arg %s", optarg);
	  putc ('\n', stderr);
	  break;
	case 'h':		/* HELP */
	  free (outputFileName);
	  free (vectorName);
	  usage (EXIT_SUCCESS);
	  break;
	case 'v':		/* VERSION */
	  version_etc (stdout, program_name, PACKAGE, VERSION,
		       "Fabio Fiorina", NULL);
	  free (outputFileName);
	  free (vectorName);
	  exit (0);
	  break;
	case 'c':		/* CHECK SYNTAX */
	  checkSyntaxOnly = 1;
	  break;
	case 'o':		/* OUTPUT */
	  outputFileName = (char *) malloc (strlen (optarg) + 1);
	  strcpy (outputFileName, optarg);
	  break;
	case 'n':		/* VECTOR NAME */
	  vectorName = (char *) malloc (strlen (optarg) + 1);
	  strcpy (vectorName, optarg);
	  break;
	case '?':		/* UNKNOW OPTION */
	  fprintf (stderr,
		   "asn1Parser: option '%s' not recognized or without argument.\n\n",
		   argv[optind - 1]);
	  free (outputFileName);
	  free (vectorName);
	  usage (EXIT_FAILURE);
	  break;
	default:
	  fprintf (stderr,
		   "asn1Parser: ?? getopt returned character code Ox%x ??\n",
		   option_result);
	}

    }

  if (optind == argc)
    {
      free (outputFileName);
      free (vectorName);
      usage (EXIT_SUCCESS);
    }
  else
    {
      inputFileName = (char *) malloc (strlen (argv[optind]) + 1);
      strcpy (inputFileName, argv[optind]);
    }

  if (checkSyntaxOnly == 1)
    {
      parse_result =
	asn1_parser2tree (inputFileName, &pointer, errorDescription);
      asn1_delete_structure (&pointer);
    }
  else				/* C VECTOR CREATION */
    parse_result = asn1_parser2array (inputFileName,
				      outputFileName, vectorName,
				      errorDescription);

  switch (parse_result)
    {
    case ASN1_SUCCESS:
      fputs ("Done.\n", stderr);
      break;
    case ASN1_FILE_NOT_FOUND:
      fprintf (stderr, "asn1Parser: file %s was not found\n", inputFileName);
      break;
    case ASN1_SYNTAX_ERROR:
    case ASN1_IDENTIFIER_NOT_FOUND:
    case ASN1_NAME_TOO_LONG:
      fprintf (stderr, "asn1Parser: %s\n", errorDescription);
      break;
    default:
      fprintf (stderr, "libtasn1 ERROR: %s\n", asn1_strerror (parse_result));
    }


  free (inputFileName);
  free (outputFileName);
  free (vectorName);

  if (parse_result != ASN1_SUCCESS)
    exit (1);
  exit (0);
}
